File: examples/php/bind-object/bind_object.php
<?php
require_once __DIR__ . '/../../../mm-php/vendor/autoload.php';

use io\metamessage\jsonc\Jsonc;
use io\metamessage\jsonc\JsoncBinder;

class User {
    public string $name;
    public int $age;
    public bool $active;
    public array $scores;
}

// JSONC 字符串
$jsonc = '{
    // mm: type=str; desc=姓名
    "name": "Alice",
    // mm: type=i; desc=年龄
    "age": 25,
    // mm: type=bool; desc=是否激活
    "active": true,
    // mm: type=array; child_type=i; desc=分数
    "scores": [95, 87, 92]
}';

echo "Input JSONC:\n";
echo $jsonc . "\n";

// 解析 JSONC
$node = Jsonc::parseFromString($jsonc);

// 绑定到 User 对象
$user = new User();
JsoncBinder::bind($node, $user);

echo "\nBound to object:\n";
echo "Name: {$user->name}\n";
echo "Age: {$user->age}\n";
echo "Active: " . ($user->active ? "true" : "false") . "\n";
echo "Scores: " . implode(", ", $user->scores) . "\n";

File: examples/swift/bind-object/BindObject.swift
import MetaMessage

// JSONC 字符串
let jsonc = """
{

    // mm: desc=姓名
    "name": "Alice",

    // mm: desc=年龄
    "age": 25,

    // mm: desc=是否激活
    "active": true,
    
    // mm: desc=分数
    "scores": [95, 87, 92]
}
"""

print("Input JSONC:")
print(jsonc)

// 从 JSONC 编码到 Wire 格式
if let wire = try? fromJSONC(jsonc) {
    print("\nEncoded to Wire:")
    print(bytesToHex(wire))

    // 使用 valueToJSONC 转换
    let user = User(name: "Alice", age: 25, active: true, scores: [95, 87, 92])
    if let jsoncOut = try? valueToJSONC(user) {
        print("\nBound to object:")
        print(jsoncOut)
    }
}

func bytesToHex(_ data: Data) -> String {
    return data.map { String(format: "%02x", $0) }.joined()
}
